src/python/pants/backend/jvm/tasks/junit_run.py

```diff
@@ -490,7 +490,22 @@ def _iter_batches(self, test_registry):
       lambda tgt: tgt.concurrency,
       lambda tgt: tgt.threads)
 
-    for properties, tests in sorted(tests_by_properties.items()):
+    # N.B. Python 3 does not allow comparisons between None and other types like str.
+    # Several of the properties, which act as dictionary keys, may have None values, whereras a
+    # another has a non-None value for the same property, so comparison when sorting would fail.
+    # We provide this custom key function to avoid such invalid comparisons, while still allowing us
+    # to use None to represent non-present properties.
+    def _sort_properties(properties_with_tests):
+      workdir, platform, extra_jvm_options, extra_env_vars, concurrency, threads = properties_with_tests[0]
+      return (
+        workdir or '',
+        platform,
+        extra_jvm_options or [],
+        extra_env_vars or {},
+        concurrency or "",
+        threads or 0)
+
+    for properties, tests in sorted(tests_by_properties.items(), key=_sort_properties):
```

Comment:
I also considered instead defining an `__lt__()` for `JUnitTests`, but John and I agreed that it is cleaner to define this `key` function here locally in order to keep the change closer to where we need it and to not try to guess a generic `__lt__()` implementation when we don't know how it would be used in other applications.